src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_TypeChar.cs

```diff
@@ -327,12 +327,27 @@ private bool IsCommitCharacter(char ch)
             var selectedItem = GetExternallyUsableCompletionItem(model.SelectedItem);
             if (selectedItem.IsBuilder)
             {
-                return char.IsLetterOrDigit(ch);
+                return char.IsLetterOrDigit(typedChar);
             }
 
             var filterText = GetCurrentFilterText(model, selectedItem);
+            var isCommitCharacter = GetCompletionRules().IsCommitCharacter(selectedItem, typedChar, filterText);
+
+            // When a character, even appears to be a commit character, was typed at the begining of the filter span when it was empty,
+            // it's most likely that the user just realized he/she forgot to do something, pressed Ctrl+Left and typed it.
+            // i.e. blah| -> |blah -> !|blah
+            // but not: A a = new | -> A a = new (|
+            // Hence we should not commit the selected item.
+            if (isCommitCharacter && filterText.Length == 1)
```

Comment:
Instead of doing this, can you get the current filter span and just check to see if the caret is at the start position with a "commit character" as the typed char?